Handling incosistent TabularInput keys

therealdeadly opened the pull request:
This handles the inconsistent logging of TabularInput keys as follows:
 
1. Read the csv logfile using `csv.DictReader` iterable.
2. Place the old `_writer` pointer at the start `seek(0)`
3. Write the new header and all the rows 

Another approach to bypass the re-reads/writes in the above solution could have been to keep the logs in memory simply till the `dump() ` call and at the end write with the final header encompassing all the keys. Decided against this as it's more prone to data loss on instances of failures and missed `dump()` call, ignoring the memory cost in itself.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -35,19 +35,26 @@ def record(self, data, prefix=''):
-                    extrasaction='ignore')
+                self._writer = csv.DictWriter(self._log_file,
+                                              fieldnames=self._fieldnames,
+                                              extrasaction='ignore')
```
Why use `extrasaction='ignore'` here? It seems like `extrasaction='raise'` would be a safer choice.

therealdeadly replied:
I used `extrasaction='ignore'` because it was previously set to `'ignore'`. 
I figured because of the custom handling of mis-matching fieldnames making it redundant. But I do agree `'raise'` will be a safer choice and will make the change.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -35,19 +35,26 @@ def record(self, data, prefix=''):
+                    row_iterator = csv.DictReader(fp)
+
+                    self._log_file.seek(0)  # put pointer back at start
+                    self._writer = csv.DictWriter(
```
It's probably best to refactor out the writer initialization logic, so that if the parameters get changed somewhere they're consistently changed.

therealdeadly replied:
Good catch, will refactor it out.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -35,19 +35,26 @@ def record(self, data, prefix=''):
+                        extrasaction='ignore',
+                    )
+                    self._writer.writeheader()
+                    self._writer.writerows(row_iterator)  # update the prev
```
How do we know that the row_iterator has read far enough ahead of the writer for this to be correct?

therealdeadly replied:
Since `row_iterator` is inherently a [`csv.reader`](https://docs.python.org/3/library/csv.html#csv.reader) object, which as you can see in the docs:
> will iterate over lines in the given csvfile

The [`csv.DictWriter.writerows`](https://docs.python.org/3/library/csv.html#csv.csvwriter.writerows) method takes an iterable of row objects(here `row_iterator`). 

`self._writer.writerows(row_iterator)`  then simply defines a _pipe_ between `row_iterator` and `self._writer`. Where 
`writerows` iterates over  `row_iterator` which in turn iterates over csvfiles line by line, making `row_iterator` stay +1 line ahead of the writer. 
Does this clear it up? I could write a simple script displaying this behaviour if you'd like.

therealdeadly commented:
@krzentner I have updated the PR with the relevant changes, please take a look. Also I changed the unit test to be more complex and it can be easily extended further if required.

therealdeadly commented:
Reading every row into memory didn't seem efficient for large logs, which usually is the case. So now using temporary file as buffer to update the csv and then copying it over. @krzentner what do you think?
